folding-schemes/src/folding/hypernova/decider_eth.rs

```diff
@@ -0,0 +1,309 @@
+        mut rng: impl RngCore + CryptoRng,
+        prep_param: &Self::PreprocessorParam,
+        fs: FS,
+    ) -> Result<(Self::ProverParam, Self::VerifierParam), Error> {
+        let circuit =
+            DeciderEthCircuit::<C1, GC1, C2, GC2, CS1, CS2>::from_hypernova::<FC, MU, NU>(
+                fs.into(),
+            )
+            .unwrap();
+
+        // get the Groth16 specific setup for the circuit
+        let (g16_pk, g16_vk) = S::circuit_specific_setup(circuit, &mut rng).unwrap();
+
+        // get the FoldingScheme prover & verifier params from HyperNova
+        #[allow(clippy::type_complexity)]
+        let hypernova_pp: <HyperNova<C1, GC1, C2, GC2, FC, CS1, CS2, MU, NU,false> as FoldingScheme<
+            C1,
+            C2,
+            FC,
+        >>::ProverParam = prep_param.0.clone().into();
+        #[allow(clippy::type_complexity)]
+        let hypernova_vp: <HyperNova<C1, GC1, C2, GC2, FC, CS1, CS2, MU, NU,false> as FoldingScheme<
+            C1,
+            C2,
+            FC,
+        >>::VerifierParam = prep_param.1.clone().into();
+        let pp_hash = hypernova_vp.pp_hash()?;
+
+        let pp = (g16_pk, hypernova_pp.cs_pp);
+        let vp = (pp_hash, g16_vk, hypernova_vp.cs_vp);
+        Ok((pp, vp))
+    }
+
+    fn prove(
+        mut rng: impl RngCore + CryptoRng,
+        pp: Self::ProverParam,
+        folding_scheme: FS,
+    ) -> Result<Self::Proof, Error> {
+        let (snark_pk, cs_pk): (S::ProvingKey, CS1::ProverParams) = pp;
+
+        let circuit = DeciderEthCircuit::<C1, GC1, C2, GC2, CS1, CS2>::from_hypernova::<FC, MU, NU>(
+            folding_scheme.into(),
+        )?;
+
+        let snark_proof = S::prove(&snark_pk, circuit.clone(), &mut rng)
+            .map_err(|e| Error::Other(e.to_string()))?;
+
+        let rho_Fr = circuit.rho.unwrap();
+        let W_i1 = circuit.W_i1.unwrap();
+
+        // get the challenges that have been already computed when preparing the circuit inputs in
+        // the above `from_hypernova` call
+        let challenge_W = circuit
+            .kzg_challenge
+            .ok_or(Error::MissingValue("kzg_c_W".to_string()))?;
+
+        // generate KZG proofs
+        let U_cmW_proof = CS1::prove_with_challenge(
+            &cs_pk,
+            challenge_W,
+            &W_i1.w,
+            &C1::ScalarField::zero(),
+            None,
+        )?;
+
+        Ok(Self::Proof {
+            snark_proof,
+            kzg_proof: U_cmW_proof,
+            rho: rho_Fr,
+            U_i1: circuit.U_i1.unwrap(),
+            kzg_challenge: challenge_W,
+        })
+    }
+
+    fn verify(
+        vp: Self::VerifierParam,
+        i: C1::ScalarField,
+        z_0: Vec<C1::ScalarField>,
+        z_i: Vec<C1::ScalarField>,
+        // we don't use the instances at the verifier level, since we check them in-circuit
+        _running_instance: &Self::CommittedInstance,
+        _incoming_instance: &Self::CommittedInstance,
+        proof: &Self::Proof,
```

Comment:
proof will be indeed consumed. So it makes sense to just pass it by value instead.

Reply:
oh, ok. Rust question (from my ignorance): we are only cloning internally 3 params of the Proof struct, but the majority of the struct parameters are passed by reference to other methods (not-cloning), it's still worth to pass it by value then? or we're getting some benefit at the places where it is still by reference?
(asking from total ignorance)

Reply:
The destructor will be called once the scope where `proof` lives ends.
Therefore, if this scope is open for a lot more time, we are holding a reference and making it live longer for no real reason.

Also, it's a lot easier to see that after calling verify, the input is no longer needed if we pass the element by value.
And also makes the user think that if needs to clone always when sending the value inside, might be doing something wrong.

I have to say that's not Rust guidelines. Rather taste. But I think definitely makes sense to pass by value something you know you'll not use anymore during the rest of the outer fn call scope.

Reply:
Oh I found a different case, where the `proof` struct is still used after calling `verify`. For the ethereum case, we want to first call `Decider::verify(..., proof)`, and then once we know that passes the verification we want to call the `prepare_calldata(..., proof)`.
If we remove the pass-by-reference, then we need to do
```rust
Decider::verify(..., proof.clone());
prepare_calldata(..., proof);
```
whereas with pass-by-reference would be:
```rust
Decider::verify(..., &proof);
prepare_calldata(..., &proof);
```


Another case where this will happen is in the normal flow where the prover is in one device, and wants to first use Decider::verify, and then once is sure that it passes the verification, wants to serialize the Proof into bytes to send it over the internet to the verifier.

Does with this make sense to pass by reference? or it still makes sense to do it by value and then do the `.clone()` on the verify call? I'm inclined towards keeping with pass-by-reference now.

Reply:
Cool. Not only for the solidity verification use case, but also in the normal use case, usually the prover will want to run the verification (using the `&proof`) and once sure about the proof passing the verification, the prover will want to serialize the proof into bytes to then send it over the internet to the verifier.
Will leave it as it was then.